src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,77 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        List<Wheel> newWheels = new ArrayList<>();
+        for (Wheel currentWheel : wheels) {
+            newWheels.add(currentWheel.clone());
+        }
+        this.wheels = newWheels;
+        this.engine = engine == null ? null : engine.clone();
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, getWheels(), engine);
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(year, newColor, getWheels(), engine);
+    }
+
+    public Car addWheel(Wheel newWheel) {
+        List<Wheel> newWheels = getWheels();
+        newWheels.add(newWheel.clone());
+        return new Car(year, color, newWheels, engine);
```

Comment:
```suggestion
        Car newCar = new Car(year, color, newWheels, engine);
        newCar.wheels.add(newWheel.clone());
        return newCar;
```
this way you won't clone wheels twice